Update to Theia version 1.40.1

sgraband opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
**Merge Theia Blueprint update to 1.40.0**
- Updates CDT.cloud Blueprint to commit https://github.com/eclipsesource/cdt-cloud-blueprint/commit/66a403276a1866f0196ee130b6f697e85f085e59 of Theia Blueprint.

**Update Theia versions to 1.40.0**
- Update docker application to resemble browser app
- Fix lerna scope expressions (blueprint -> cdt-cloud-blueprint)
- Also build @eclipse-cdt-cloud/ extensions
- Move plugins to root and remove from applications
- Update theia-traceviewer to 0.2.0-next.20230908184630.abe6fad.0
- Fix dockerfile
- Update inversify to ^6.0.1
- Update yarn.lock

**Update to 1.40.1**

**Update theia-traceviewer to newest version**

- Update to 0.2.0-next.20230908184630.abe6fad.0.
- This now removed to requirement to expose the 8080 port.
- Also update the next version of tsp-typescript-client.

Contributed on behalf of STMicroelectronics

jfaltermeier committed:
Update to 1.39.0

* update gitignore
* update path to electron-main script
* upgrade dependencies compatible with node 16
* inversify update changes

Contributed on behalf of STMicroelectronics

Signed-off-by: Johannes Faltermeier <jfaltermeier@eclipsesource.com>

marcdumais-work committed:
[build][Jenkins] more efficient Jenkins CI, builds, improved linter config

This is a significant refactoring of the repo and CI.

1) Jenkins CI: more efficient, quicker builds

I propose that we adopt a "run often, run quickly" approach to our Jenkins CI.
ATM we mostly run Jenkins when it's time to release new Blueprint installers
(when a PR is merged on master). The other case is if "jenkins" is detected
in the PR branch or PR title, then almost a full Jenkins build is done, just
stopping short of publishing new Blueprint installers (full signing and
notarization is done, which can take a couple of hours).

I made it so we run a very light version of the Jenkins CI by default, that
builds the "dev" version of the app and pre-packages it using electron builder,
generating a directory that contains the app but no installers.

We really ought to run tests then, on the unpacked pre-packaged app, but
it will need to be done separately. Until then, we do not require the built-in
extensions (plugins), so I gave openvsx.org a rest and skip fetching them, by
default.

When it's detected that the Jenkins CI is processing the result of a merge on
master or if the new "release dry-run" mode is enabled [1], a production version
of the Blueprint apps will be produced, including built-in plugins, which will
then be packaged for "real", producing genuine OS-specific installers, signed
and notarized as required. This will take a lot longer to run, of course, but
may be required only a minority of the time.

Conclusion: with this PR, running Jenkins CI now takes about 12 minutes, and
since this is relatively quick, we can afford to run this when any PR branch
is created or updated.

Running the same CI in "release dry-run mode" takes about an hour. A release
will take longer, a bit over an hour, since additional steps are executed,
like copying the various installers to a release folder.

[1] About the "release dry-run" mode, it can be enabled in JenkinsFile,
"pipeline" definition near the top, in "environment":
  BLUEPRINT_JENKINS_RELEASE_DRYRUN = 'true'

I wish I found a better mechanism to enable this "release dry-run" mode. But
this will serve for now. The "release dry-run" mode should be disabled before
merging a PR to master, but can in the meantime be used to test or troubleshoot
the whole pipeline, in particular signing and notarizing.

2) tsconfig/eslint configs:

  Added configs, to improve linter coverage. This made it possible for some
  source files, not previously covered, to get ts/linter feedback, both while
  editing and when running `yarn lint`. This will help keep code in-line with
  our standards. The config is not perfect and I would welcome further
  improvements. But for now I think it's a nice improvement.

3) Build "scripts" (package.json)

Refactored the build commands ("scripts" section in package.json)

- previously, merely running 'yarn" in the repo's root would rebuild
  every application from scratch. This prevents running a quick
  "yarn install", e.g. just to re-install build dependencies
- the new version permits a granular build, with simple defaults
- inspired from a similar change not so long ago in the main repo
- see updated README for some examples

Other misc items:

- renamed extensions / applications
- made the browser application a first class citizen, equal to the
  Electron application
- all applications now share a common 'plugins' folder rather than
  each having their own. Moved the plugin-related entries to root
  package.json
- to gain flexibility about which `yarn workspaces` are invoked for a
  given `lerna` command, using the `--scope=` CLI option.I renamed the
  repo's extensions and applications. This permits easily composing
  commands that target only the extensions or only the applications.
  e.g.:

  ``` json
  "build:extensions": "lerna run --scope=\"blueprint*ext\" build",
  "build:applications": "lerna run --scope=\"blueprint*app\" build --concurrency 1","
  ```

- renamed the extensions folders, made them more straightforward
- For systems with limited RAM, like on a Raspberry Pi 4B board with 4GB of RAM,
  it's now possible to successfully build Blueprint. e.g. use the following cmd:
  `yarn && yarn build:dev`, optionally followed by a packing command like:
  `yarn electron package`
  - (build:dev will build the Blueprint app in dev mode, which can be achieved
    in 4GB RAM)
- [windows][jenkins] stash only dist folder: Currently, and only for Windows,
  we stash the whole git repo, which is very big and takes long to stash and
  un-stash. For the other OS, we only stash the dist folder, that contains the
  platform-specific installer, that we built. Let's try that for Windows too,
  and see if it works.
- [jenkins] Decrease timeout from 5h to 3h" Looking at the build history, all
  recent builds that succeeded, did do under 2h30. OTOH, when a build hangs,
  it needs to wait for the timeout to expire, wasting time at the current value
  of 5h. Let's compromise at 3 hours and see how it goes?
- [jenkins][installer build] exclude browser app for now: We do not yet publish
  the browser app, so let's skip building it to save time/resources. We may
  revisit when we use the new browser app bundling, recently made available
  upstream in the Theia framework.

Signed-off-by: Marc Dumais <marc.dumais@ericsson.com>

jfaltermeier committed:
Add secondary window to electron app

Contributed on behalf of STMicroelectronics

Signed-off-by: Johannes Faltermeier <jfaltermeier@eclipsesource.com>

msujew committed:
Enable backend bundling

* adjust electron-builder config and Dockerfile

Co-authored-by: Johannes Faltermeier <jfaltermeier@eclipsesource.com>
Signed-off-by: Johannes Faltermeier <jfaltermeier@eclipsesource.com>

jfaltermeier committed:
Fix Metadata Update on Windows

jfaltermeier committed:
Remove timers import from node and use from Window instead #273

Contributed on behalf of STMicroelectronics

Signed-off-by: Johannes Faltermeier <jfaltermeier@eclipsesource.com>

jfaltermeier committed:
Update to Theia 1.40.0

* upgrade dependencies as well
* fix test case by moving mouse over header, and wait until clickable.
before the click just selected the whole extensions view for some
reason.

jfaltermeier committed:
Fix next build

marcdumais-work committed:
[prod build] add rebuild

https://github.com/eclipse-theia/theia-blueprint/pull/272#pullrequestreview-1539121864

Signed-off-by: Marc Dumais <marc.dumais@ericsson.com>

marcdumais-work committed:
[browser.dockerfile] Fixed build issue

Oops - I had forgotten to build the Theia extensions, which caused a
problem at runtime.

Signed-off-by: Marc Dumais <marc.dumais@ericsson.com>

danepowell committed:
Remove scm-extra

marcdumais-work committed:
[Jenkins] Follow-up: fix detection of production mac build

To save build time, we only proceed with macos signing/notarizing when it's
detected that the current build is producing official blueprint packages.
Unfortunately a little bug prevents this from working. This is the fix.

Signed-off-by: Marc Dumais <marc.dumais@ericsson.com>

danepowell committed:
remove from electron

jfaltermeier committed:
Fix next build

jfaltermeier committed:
No app icon for AppImage on Linux #296

* consume image from applications/electron/resources/icons/WindowIcon/

Contributed on behalf of STMicroelectronics

Signed-off-by: Johannes Faltermeier <jfaltermeier@eclipsesource.com>

sgraband committed:
Merge Theia Blueprint update to 1.40.0

Updates CDT.cloud Blueprint to commit 66a4032 of Theia Blueprint.

Contributed on behalf of STMicroelectronics

sgraband committed:
Update Theia versions to 1.40.0

- Update docker application to resemble browser app
- Fix lerna scope expressions (blueprint -> cdt-cloud-blueprint)
- Also build @eclipse-cdt-cloud/ extensions
- Move plugins to root and remove from applications
- Update theia-traceviewer to 0.2.0-next.20230908184630.abe6fad.0
- Fix dockerfile
- Update inversify to ^6.0.1
- Update yarn.lock

Contributed on behalf of STMicroelectronics

sgraband committed:
Update to 1.40.1

Contributed on behalf of STMicroelectronics

sgraband committed:
Update theia-traceviewer to newest version

Update to 0.2.0-next.20230908184630.abe6fad.0.
This now removed to requirement to expose the 8080 port.
Also update the next version of tsp-typescript-client.

Contributed on behalf of STMicroelectronics

sgraband committed:
Fix linting issues

Contributed on behalf of STMicroelectronics

sgraband committed:
Fix tests

Contributed on behalf of STMicroelectronics

sdirix reviewed (changes requested):
The example generator does not work. I suspect that the backend bundling is responsible for this. To fix this webpack must be customized to copy the example files to the correct location, similar to [this PR](https://github.com/eclipse-theia/theia/pull/12787/files) in Theia itself.

sgraband committed:
Bundle examples with webpack

This makes sure that they are available during runtime.
Use _dirname instead of module.path as the latter doesn't work.

Contributed on behalf of STMicroelectronics
